### DISCUSSION TITLE
Trying to use dabbrev-capf within capf with company

### DISCUSSION THREAD

**Discussion by garyo** [company-mode/company-mode#1433] [Q&A] [answered]
I'm using company (with vertico etc.) for completion. The default backend is `company-capf`:
```elisp
  (setq company-backends '(company-capf company-dabbrev-code
                                        company-dabbrev company-etags
                                        company-keywords))
```
I'm trying to set things up so by default I get `dabbrev` completion. I'm trying to set that up in `capf` like this:
```elisp
;; Use ~setq-default~ because we need to set the global version, not buffer-local.
;; Otherwise we'd normally use ~add-to-list~.
(setq-default completion-at-point-functions (cons 'dabbrev-capf completion-at-point-functions))
```
I can see the capf list is OK, but when I go to a markdown mode buffer (which uses the global value of `capf`) and try to complete a word, I get this error:
```
Company: An error occurred in auto-begin
Company: backend company-capf error "Symbol’s function definition is void: dabbrev-capf" with args (prefix)
```
The full traceback is like this:
```
Debugger entered--Lisp error: (void-function dabbrev-capf)
  dabbrev-capf()
  completion--capf-wrapper(dabbrev-capf optimist)
  company--capf-wrapper(dabbrev-capf optimist)
  company--capf-data-real()
  company--capf-data()
  company-capf(prefix)
  apply(company-capf prefix)
  company-call-backend-raw(prefix)
  company--force-sync(company-call-backend-raw (prefix) company-capf)
  company-call-backend(prefix)
  company--begin-new()
  company--perform()
  company-auto-begin()
  company-manual-begin()
  company-complete()
```
Am I not setting it up right? Looks like `dabbrev-capf` doesn't take any args. Does it need to?

BTW the reason I want to add it at the `capf` level rather than in `company-backends` is because sometimes LSP modes or eglot modify things so it seems (?) like setting the default value of `capf` is a good way to ensure it's always used.

**Comment by dgutov** [comment, marked as answer]
> Looks like dabbrev-capf doesn't take any args. Does it need to?

From the backtrace, it just says like `dabbrev-capf` is not defined in your session. That probably just means that you need to `(require 'dabbrev)` in your init file somewhere (this function was added in Emacs 29, but it's not autoloaded).

But the completion function itself doesn't seem to be well-written and debugged: I've just tried it, and it shows a backtrace after a backtrace. If somebody has used it successfully (@minad?), they're welcome to chime in.

You might want to try `cape-dabbrev` as alternative, but according to my measurements (#1411) it's overall slower than the company backends (either of the two).

> BTW the reason I want to add it at the capf level rather than in company-backends is because sometimes LSP modes or eglot modify things so it seems (?) like setting the default value of capf is a good way to ensure it's always used.

I recommend instead customizing a variable like `eglot-stay-out-of` or `lsp-completion-provider` (setting the latter to `:none`). Depending on which of the LSP clients you use. And customizing `company-backends` to the preferred configuration.

**Reply by garyo** [reply]
That makes sense. I'll go that way, adding to `company-backends`; that also has the benefit of being able to separate results by backend. (Also: I did try `cape-dabbrev` but it gave weird/incorrect results.)
